pkg/api/customization/catalog/template.go

```diff
@@ -3,20 +3,68 @@ package catalog
 import (
 	"bytes"
 	"encoding/base64"
+	"fmt"
 	"net/http"
 	"strings"
 	"time"
 
+	"github.com/rancher/norman/types/convert"
+
 	"github.com/rancher/norman/api/access"
 	"github.com/rancher/norman/httperror"
 	"github.com/rancher/norman/types"
 	"github.com/rancher/rancher/pkg/templatecontent"
 	v3 "github.com/rancher/types/apis/management.cattle.io/v3"
 	managementschema "github.com/rancher/types/apis/management.cattle.io/v3/schema"
 	client "github.com/rancher/types/client/management/v3"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-func TemplateFormatter(apiContext *types.APIContext, resource *types.RawResource) {
+func checkUpgrade(t TemplateWrapper, templateVersionName string) (string, error) {
+	templateVersion, err := t.TemplateVersion.Get(templateVersionName, metav1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+	id := strings.Split(templateVersion.Name, "-")[1]
+	tag := templateVersion.Spec.Files[id+"/Chart.yaml"]
+	data, err := templatecontent.GetTemplateFromTag(tag, t.TemplateContentClient)
+	if err != nil {
+		return "", err
+	}
+	resp := findIconURL(data)
+	if resp == "" {
+		return "", fmt.Errorf("icon not found")
+	}
+	return resp, nil
+}
+
+func findIconURL(s string) string {
+	i := strings.Index(s, "icon:")
+	if i < 0 {
+		return ""
+	}
+	s = s[i:]
+	target := strings.Split(s, "\n")[0]
+	i = strings.Index(target, "http")
+	if i < 0 {
+		return ""
+	}
+	target = target[i:]
+	return target
+}
+
+// indexInterfaceSlice takes a list of interfaces, returns the first, asserts its a map[string]interface{}
+// and uses the provided index to retrieve the value
+func indexInterfaceSlice(value interface{}, key string) string {
```

Comment:
looks like we could have some danger of panicking at various points in the function if you get unexpected data, do you want to add some safety checks just so we dont have to worry about that?
like, i think if you simply change `v2 := v.(map[string]interface{})` to `v2, ok := v.(map[string]interface{})`, you prevent that type conversion from panicking

Reply:
Changed type assertion to test if the underlying type exists